File: Head_First_Design_Patterns/src/facade/python/HomeTheaterFacade.py
# -*- coding: utf-8 -*-
"""
Created on Wed Oct 24 18:01:22 2018

@author: 凯风
"""

class HomeTheaterFacade(object):
    def __init__(self, amplifier, tuner, dvdPlayer, cdPlayer, projector, screen, lights, popper):
        self.amplifier = amplifier
        self.tuner = tuner
        self.dvdPlayer = dvdPlayer
        self.cdPlayer = cdPlayer
        self.projector = projector
        self.screen = screen
        self.lights = lights
        self.popper = popper
    
    def watchMovie(self, movie):
        print("Get ready to watch a movie...")
        self.popper.on()
        self.popper.pop()
        self.lights.dim(10)
        self.screen.down()
        self.projector.on()
        self.projector.wideScreenMode()
        self.amplifier.on()
        self.amplifier.setDvd(self.dvdPlayer)
        self.amplifier.setSurroundSound()
        self.amplifier.setVolume(15)
        self.dvdPlayer.on()
        self.dvdPlayer.play(movie = movie)
    
    def endMovie(self):
        print("Shutting movie theater down...")
        self.popper.off()
        self.lights.on()
        self.screen.up()
        self.projector.off()
        self.amplifier.off()
        self.dvdPlayer.stop()
        self.dvdPlayer.eject()
        self.dvdPlayer.off()
    
    def listenToCd(self, cdTitle):
        print("Get ready for an audiopile experence...")
        self.lights.on()
        self.amplifier.on()
        self.amplifier.setVolume(5)
        self.amplifier.setCd(self.cdPlayer)
        self.amplifier.setStereoSound()
        self.cdPlayer.on()
        self.cdPlayer.play(title = cdTitle)
    
    def endCd(self):
        print("Shutting down CD...")
        self.amplifier.off()
        self.amplifier.setCd(self.cdPlayer)
        self.cdPlayer.eject()
        self.cdPlayer.off()
    
    def listenToRadio(self, frequency):
        print("Tuning in the airwaves...")
        self.tuner.on()
        self.tuner.setFrequency(frequency)
        self.amplifier.on()
        self.amplifier.setVolume(7)
        self.amplifier.setTuner(self.tuner)
    
    def endRadio(self):
        print("Shutting down the tuner...")
        self.tuner.off()
        self.amplifier.off()
                        
    

File: Head_First_Design_Patterns/src/facade/HomeTheaterFacade.java
package facade;

public class HomeTheaterFacade {
    Amplifier amplifier;
    Tuner tuner;
    DvdPlayer dvdPlayer;
    CdPlayer cdPlayer;
    Projector projector;
    TheaterLights lights;
    Screen screen;
    PopcornPopper popper;

    public HomeTheaterFacade(Amplifier amplifier,
                             Tuner tuner,
                             DvdPlayer dvdPlayer,
                             CdPlayer cdPlayer,
                             Projector projector,
                             Screen screen,
                             TheaterLights lights,
                             PopcornPopper popper){
        this.amplifier = amplifier;
        this.tuner = tuner;
        this.dvdPlayer = dvdPlayer;
        this.cdPlayer = cdPlayer;
        this.projector = projector;
        this.lights = lights;
        this.screen = screen;
        this.popper = popper;
    }

    public void watchMovie(String movie){
        System.out.println("Get ready to watch a movie...");
        popper.on();
        popper.pop();
        lights.dim(10);
        screen.down();
        projector.on();
        projector.wideScreenMode();
        amplifier.on();
        amplifier.setDvd(dvdPlayer);
        amplifier.setSurroundSound();
        amplifier.setVolume(15);
        dvdPlayer.on();
        dvdPlayer.play(movie);
    }

    public void endMovie(){
        System.out.println("Shutting movie theater down...");
        popper.off();
        lights.on();
        screen.up();
        projector.off();
        amplifier.off();
        dvdPlayer.stop();
        dvdPlayer.eject();
        dvdPlayer.off();
    }

    public void listenToCd(String cdTitle){
        System.out.println("Get ready for an audiopile experence...");
        lights.on();
        amplifier.on();
        amplifier.setVolume(5);
        amplifier.setCd(cdPlayer);
        amplifier.setStereoSound();
        cdPlayer.on();
        cdPlayer.play(cdTitle);
    }

    public void endCd(){
        System.out.println("Shutting down CD...");
        amplifier.off();
        amplifier.setCd(cdPlayer);
        cdPlayer.eject();
        cdPlayer.off();
    }

    public void listenToRadio(double frequency){
        System.out.println("Tuning in the airwaves...");
        tuner.on();
        tuner.setFrequency(frequency);
        amplifier.on();
        amplifier.setVolume(7);
        amplifier.setTuner(tuner);
    }

    public void endRadio(){
        System.out.println("Shutting down the tuner...");
        tuner.off();
        amplifier.off();
    }
}
